File: frontend/user-frontend/src/app/core/services/task.service.ts
import { Injectable, inject } from '@angular/core';
import { Apollo, gql } from 'apollo-angular';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import {
  Task,
  TaskPage,
  CreateTaskInput,
  UpdateTaskStatusInput,
  AssignTaskInput,
  AddCommentInput,
  TaskFilter
} from '../models/task.model';

const GET_TASKS = gql`
  query GetTasks($filter: TaskFilter, $page: Int, $size: Int) {
    tasks(filter: $filter, page: $page, size: $size) {
      content {
        id
        title
        description
        status
        priority
        dueDate
        createdAt
        updatedAt
        assignedTo
        createdBy
        tags
      }
      totalElements
      totalPages
      pageNumber
      pageSize
    }
  }
`;

const GET_MY_TASKS = gql`
  query GetMyTasks($page: Int, $size: Int) {
    myTasks(page: $page, size: $size) {
      content {
        id
        title
        description
        status
        priority
        dueDate
        createdAt
        updatedAt
        assignedTo
        createdBy
        tags
      }
      totalElements
      totalPages
      pageNumber
      pageSize
    }
  }
`;

const GET_TASK = gql`
  query GetTask($id: ID!) {
    task(id: $id) {
      id
      title
      description
      status
      priority
      dueDate
      createdAt
      updatedAt
      assignedTo
      createdBy
      tags
      comments {
        id
        content
        createdAt
        userId
      }
    }
  }
`;

const CREATE_TASK = gql`
  mutation CreateTask($input: CreateTaskInput!) {
    createTask(input: $input) {
      id
      title
      description
      status
      priority
      dueDate
      createdAt
      assignedTo
      createdBy
    }
  }
`;

const UPDATE_TASK_STATUS = gql`
  mutation UpdateTaskStatus($id: ID!, $status: TaskStatus!) {
    updateTaskStatus(id: $id, status: $status) {
      id
      status
      updatedAt
    }
  }
`;

const ASSIGN_TASK = gql`
  mutation AssignTask($id: ID!, $userId: ID!) {
    assignTask(id: $id, userId: $userId) {
      id
      assignedTo
    }
  }
`;

const ADD_COMMENT = gql`
  mutation AddComment($taskId: ID!, $content: String!) {
    addComment(taskId: $taskId, content: $content) {
      id
      content
      createdAt
      userId
    }
  }
`;

@Injectable({ providedIn: 'root' })
export class TaskService {
  private apollo = inject(Apollo);

  getTasks(filter?: TaskFilter): Observable<TaskPage> {
    return this.apollo.watchQuery<{ tasks: TaskPage }>({
      query: GET_TASKS,
      variables: {
        filter: filter ? { status: filter.status, priority: filter.priority, assignedTo: filter.assigneeId } : null,
        page: filter?.page ?? 0,
        size: filter?.size ?? 20,
      }
    }).valueChanges.pipe(
      map(result => result.data.tasks)
    );
  }

  getMyTasks(status?: string, page = 0, size = 20): Observable<TaskPage> {
    return this.apollo.watchQuery<{ myTasks: TaskPage }>({
      query: GET_MY_TASKS,
      variables: { page, size }
    }).valueChanges.pipe(
      map(result => result.data.myTasks)
    );
  }

  getTask(id: string): Observable<Task> {
    return this.apollo.watchQuery<{ task: Task }>({
      query: GET_TASK,
      variables: { id }
    }).valueChanges.pipe(
      map(result => result.data.task)
    );
  }

  createTask(input: CreateTaskInput): Observable<Task> {
    return this.apollo.mutate<{ createTask: Task }>({
      mutation: CREATE_TASK,
      variables: { input },
      refetchQueries: [{ query: GET_TASKS }, { query: GET_MY_TASKS }]
    }).pipe(
      map(result => result.data!.createTask)
    );
  }

  updateTaskStatus(input: UpdateTaskStatusInput): Observable<Partial<Task>> {
    return this.apollo.mutate<{ updateTaskStatus: Partial<Task> }>({
      mutation: UPDATE_TASK_STATUS,
      variables: { id: input.taskId, status: input.status }
    }).pipe(
      map(result => result.data!.updateTaskStatus)
    );
  }

  assignTask(input: AssignTaskInput): Observable<Partial<Task>> {
    return this.apollo.mutate<{ assignTask: Partial<Task> }>({
      mutation: ASSIGN_TASK,
      variables: { id: input.taskId, userId: input.assigneeId }
    }).pipe(
      map(result => result.data!.assignTask)
    );
  }

  addComment(input: AddCommentInput): Observable<any> {
    return this.apollo.mutate({
      mutation: ADD_COMMENT,
      variables: { taskId: input.taskId, content: input.content }
    }).pipe(
      map(result => (result.data as any).addComment)
    );
  }
}

File: backend/task-service/src/main/kotlin/com/taskmaster/task/task/TaskService.kt
package com.taskmaster.task.task

import com.fasterxml.jackson.databind.ObjectMapper
import com.taskmaster.task.comment.TaskCommentRepository
import com.taskmaster.task.history.TaskHistoryService
import com.taskmaster.task.outbox.OutboxEvent
import com.taskmaster.task.outbox.OutboxEventRepository
import com.taskmaster.task.task.dto.CreateTaskInput
import com.taskmaster.task.task.dto.TaskDto
import com.taskmaster.task.task.dto.TaskFilter
import com.taskmaster.task.task.dto.TaskPage
import com.taskmaster.task.task.dto.UpdateTaskInput
import org.slf4j.LoggerFactory
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.util.UUID

@Service
@Transactional
class TaskService(
    private val taskRepository: TaskRepository,
    private val outboxRepository: OutboxEventRepository,
    private val objectMapper: ObjectMapper,
    private val taskHistoryService: TaskHistoryService,
    private val taskCommentRepository: TaskCommentRepository
) {
    private val log = LoggerFactory.getLogger(TaskService::class.java)

    @Transactional(readOnly = true)
    fun findAll(tenantId: UUID, filter: TaskFilter?, page: Int, size: Int): TaskPage {
        val spec = TaskSpecification.fromFilter(tenantId, filter)
        val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "createdAt"))
        val result = taskRepository.findAll(spec, pageable)
        return TaskPage(
            content = result.content.map { it.toDtoWithComments() },
            totalElements = result.totalElements,
            totalPages = result.totalPages,
            pageNumber = result.number,
            pageSize = result.size
        )
    }

    @Transactional(readOnly = true)
    fun findById(id: UUID, tenantId: UUID): TaskDto? {
        return taskRepository.findByIdAndTenantId(id, tenantId)?.toDtoWithComments()
    }

    @Transactional(readOnly = true)
    fun findByAssignee(tenantId: UUID, assignedTo: UUID, page: Int, size: Int): TaskPage {
        val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "createdAt"))
        val result = taskRepository.findByAssignedToAndTenantId(assignedTo, tenantId, pageable)
        return TaskPage(
            content = result.content.map { it.toDtoWithComments() },
            totalElements = result.totalElements,
            totalPages = result.totalPages,
            pageNumber = result.number,
            pageSize = result.size
        )
    }

    fun createTask(tenantId: UUID, createdBy: UUID, input: CreateTaskInput): TaskDto {
        val task = Task(
            title = input.title,
            description = input.description,
            status = TaskStatus.TODO,
            priority = input.priority ?: TaskPriority.MEDIUM,
            dueDate = input.dueDate,
            tenantId = tenantId,
            createdBy = createdBy,
            assignedTo = input.assignedTo,
            tags = input.tags?.toMutableSet() ?: mutableSetOf()
        )
        val saved = taskRepository.save(task)

        input.assignedTo?.let { assignedTo ->
            publishOutboxEvent(saved, "TASK_ASSIGNED", assignedTo)
        }

        log.info("Task created: id={}, tenantId={}", saved.id, tenantId)
        return saved.toDto()
    }

    fun updateTask(id: UUID, tenantId: UUID, input: UpdateTaskInput): TaskDto {
        val task = taskRepository.findByIdAndTenantId(id, tenantId)
            ?: throw NoSuchElementException("Task not found: $id")

        input.title?.let { task.title = it }
        input.description?.let { task.description = it }
        input.priority?.let { task.priority = it }
        input.dueDate?.let { task.dueDate = it }
        input.tags?.let { task.tags = it.toMutableSet() }
        task.updatedAt = Instant.now()

        val saved = taskRepository.save(task)
        log.info("Task updated: id={}", id)
        return saved.toDto()
    }

    fun updateTaskStatus(id: UUID, tenantId: UUID, newStatus: TaskStatus, changedBy: UUID): TaskDto {
        val task = taskRepository.findByIdAndTenantId(id, tenantId)
            ?: throw NoSuchElementException("Task not found: $id")
        val oldStatus = task.status
        task.status = newStatus
        task.updatedAt = Instant.now()
        val saved = taskRepository.save(task)

        taskHistoryService.record(id, changedBy, oldStatus, newStatus)

        task.assignedTo?.let { assignedTo ->
            publishOutboxEvent(saved, "TASK_UPDATED", assignedTo)
        }

        log.info("Task status updated: id={}, {} -> {}", id, oldStatus, newStatus)
        return saved.toDto()
    }

    fun assignTask(id: UUID, tenantId: UUID, assignedTo: UUID, assignedBy: UUID): TaskDto {
        val task = taskRepository.findByIdAndTenantId(id, tenantId)
            ?: throw NoSuchElementException("Task not found: $id")
        task.assignedTo = assignedTo
        task.updatedAt = Instant.now()
        val saved = taskRepository.save(task)

        publishOutboxEvent(saved, "TASK_ASSIGNED", assignedTo)

        log.info("Task assigned: id={}, assignedTo={}", id, assignedTo)
        return saved.toDto()
    }

    fun deleteTask(id: UUID, tenantId: UUID) {
        val task = taskRepository.findByIdAndTenantId(id, tenantId)
            ?: throw NoSuchElementException("Task not found: $id")
        taskRepository.delete(task)
        log.info("Task deleted: id={}", id)
    }

    private fun publishOutboxEvent(task: Task, eventType: String, assignedTo: UUID) {
        val payload = objectMapper.writeValueAsString(
            mapOf(
                "eventId" to UUID.randomUUID().toString(),
                "taskId" to task.id.toString(),
                "taskTitle" to task.title,
                "tenantId" to task.tenantId.toString(),
                "assignedTo" to assignedTo.toString(),
                "eventType" to eventType,
                "timestamp" to Instant.now().toString()
            )
        )
        val idempotencyKey = "${task.id}:${eventType}:${Instant.now().epochSecond / 60}"

        try {
            outboxRepository.save(
                OutboxEvent(
                    aggregateId = task.id,
                    eventType = eventType,
                    payload = payload,
                    idempotencyKey = idempotencyKey
                )
            )
        } catch (ex: Exception) {
            log.warn("Duplicate outbox event skipped: {}", idempotencyKey)
        }
    }

    private fun Task.toDto() = TaskDto(
        id = id,
        title = title,
        description = description,
        status = status,
        priority = priority,
        dueDate = dueDate,
        scheduledDate = scheduledDate,
        tenantId = tenantId,
        assignedTo = assignedTo,
        createdBy = createdBy,
        tags = tags.toSet(),
        createdAt = createdAt,
        updatedAt = updatedAt
    )

    private fun Task.toDtoWithComments(): TaskDto {
        val comments = taskCommentRepository.findByTaskIdOrderByCreatedAtAsc(id)
            .map { c ->
                com.taskmaster.task.comment.dto.TaskCommentDto(
                    id = c.id,
                    taskId = c.taskId,
                    userId = c.userId,
                    content = c.content,
                    createdAt = c.createdAt
                )
            }
        return toDto().copy(comments = comments)
    }
}
